src/React/Hook.purs

```diff
@@ -0,0 +1,310 @@
+module React.Hook
+  ( Hook
+
+  , useState
+  , useStateLazy
+  , setState
+  , modifyState
+  , SetState
+
+  , useEffect
+  , effectInput
+  , EffectInput
+
+  , useContext
+  , Context
+
+  , useReducer
+  , useReducerLazy
+  , dispatch
+  , Dispatch
+
+  , useCallback
+  , callbackInput
+  , CallbackInput
+
+  , useMemo
+  , memoInput
+  , MemoInput
+
+  , useRef
+  , getRef
+  , setRef
+  , Ref
+
+  , useImperativeMethods
+  , imperativeMethodsInput
+  , ImperativeMethodsInput
+
+  , useMutationEffect
+  , mutationEffectInput
+  , MutationEffectInput
+
+  , useLayoutEffect
+  , layoutEffectInput
+  , LayoutEffectInput
+  ) where
+
+import Prelude
+
+import Data.Function.Uncurried (Fn1, Fn2, Fn3, Fn4, mkFn2, runFn1, runFn2, runFn3, runFn4)
+import Data.Maybe (Maybe)
+import Data.Nullable (Nullable)
+import Data.Nullable as Nullable
+import Data.Tuple (Tuple(..))
+
+import Effect (Effect)
+import Effect.Uncurried (EffectFn1, EffectFn2, runEffectFn1, runEffectFn2)
+
+import Unsafe.Coerce (unsafeCoerce)
+
+useState
+  :: forall a
+   . a
+  -> Hook (Tuple a (SetState a))
+useState = runFn2 useState_ Tuple
+
+useStateLazy
+  :: forall a
+   . (Unit -> a)
+  -> Hook (Tuple a (SetState a))
+useStateLazy = runFn2 useState_ Tuple
+
+setState
+  :: forall a
+   . SetState a
+  -> a
+  -> Effect Unit
+setState k = runEffectFn1 k'
+  where
+  k' :: EffectFn1 a Unit
+  k' = unsafeCoerce k
+
+modifyState
+  :: forall a
+   . SetState a
+  -> (a -> a)
+  -> Effect Unit
+modifyState k = runEffectFn1 k'
+  where
+  k' :: EffectFn1 (a -> a) Unit
+  k' = unsafeCoerce k
+
+foreign import data SetState :: Type -> Type
+
+foreign import useState_
+  :: forall a b
+   . Fn2 (b -> SetState b -> Tuple b (SetState b))
+         a
+         (Hook (Tuple b (SetState b)))
+
+useEffect
+  :: forall a
+   . Effect (Effect a)
+  -> Maybe (Array EffectInput)
+  -> Hook Unit
+useEffect k = runFn2 useEffect_ k <<< Nullable.toNullable
+
+effectInput :: forall a. a -> EffectInput
+effectInput = unsafeCoerce
+
+foreign import data EffectInput :: Type
+
+foreign import useEffect_
+  :: forall a
+   . Fn2 (Effect (Effect a))
+         (Nullable (Array EffectInput))
+         (Hook Unit)
+
+useContext :: forall a. Context a -> Hook a
+useContext = runFn1 useContext_
```

Comment:
Right now, this isn't usable since `Context` is a foreign type here.

Reply:
In hindsight, I should have made `Context` a foreign type to begin with instead of wrapping it.

Reply:
Maybe it is worth making `Context` a foreign type? It's a breaking change, but maybe it is an idea.

Reply:
I'm OK with the breaking change. It's very minor.

Reply:
Sounds good. I've changed `Context` over to a foreign type.

Also, in order to render the consumer, I found I needed a [createRenderPropsElement ](https://github.com/purescript-contrib/purescript-react/pull/159/files#diff-aa760d4652ab9f7e0c90ffcedd920af4R474) function. Was there another way to render the consumer?

On a side note, it seems refs have also been updated in React 16.3. I've [incorporated those changes](https://github.com/purescript-contrib/purescript-react/pull/159/commits/ebe5f47220de9c237d2b6c953f6a00d9aad56a00) as well since it fits in with the hooks API.

Reply:
I've just used `createLeafElement`, and supplied the `children` prop explicitly in those cases.

Reply:
Ah, okay. Having `createRenderPropsElement` might be useful. What do you think? I am not sure about the name though.